### PR TITLE
kvutils|ledger: Don't use Scala's global execution context.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
I have just discovered that Scala's global execution context uses daemon threads, which means they don't stick around past the end of the `main` function. This is bad when application startup fails and we end up with no other threads; the application terminates before it has a chance to print the error message, which is very unhelpful.

We resolve this by creating a cached thread pool in the `ProgramResource` class, specifically used for acquisition and release of resources.

To do this, we need to pass the execution context through to the resource constructors, which makes usage a little more verbose, because now we can't use lambda functions (which can't take implicit arguments such as the `executionContext`).

Discovered by introducing a bug into migrations which caused _ledger-on-sql_ to crash. It didn't display an error, just stopped; turns out that the program was terminating before Logback had a chance to serialize and print the stack trace.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/ProgramResource.scala [hunk 1]
```diff
@@ -3,44 +3,61 @@
...
 
-class ProgramResource[T](owner: ResourceOwner[T]) {
+class ProgramResource[T](
+    owner: ResourceOwner[T],
+    startupTimeout: FiniteDuration = 1.minute,
+    tearDownDuration: FiniteDuration = 10.seconds,
+) {
   private val logger = ContextualizedLogger.get(getClass)
 
-  def run()(implicit executionContext: ExecutionContext): Unit = {
```
Comment (stefanobaghino-da) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/ProgramResource.scala]: I believe this was a better approach. Why the change?

Comment (SamirTalwar-DA) [reply libs-scala/resources/src/main/scala/com/digitalasset/resources/ProgramResource.scala]: Hmmm, maybe you're right. Currently it's gonna be the same everywhere though. Do you think we should move it back out now or wait until something needs a different implementation?

Comment (stefanobaghino-da) [reply libs-scala/resources/src/main/scala/com/digitalasset/resources/ProgramResource.scala]: No, your argument makes sense. I still thought it made sense to raise this.
